src/parse.rs

```diff
@@ -2363,50 +2387,61 @@ impl<'a> Parser<'a> {
                                 link_stack.clear();
                                 continue;
                             }
-                            // TODO(performance): make sure we aren't doing unnecessary allocations
-                            // for the label
-                            else if let Some(matching_def) = label.and_then(|l| match l { ReferenceLabel::Link(l) => Some(l), _ => None, })
-                                .and_then(|l| self.refdefs.get(&UniCase::new(l))) {
-                                // found a matching definition!
-                                let title = matching_def.title.as_ref().cloned().unwrap_or("".into());
-                                let url = matching_def.dest.clone();
-                                self.tree[tos.node].item.body = if tos.ty == LinkStackTy::Image {
-                                    ItemBody::Image(link_type, url, title)
+                            else if let Some(ReferenceLabel::Link(link_label)) = label {
+                                let type_url_title = if let Some(matching_def) = self.refdefs.get(&UniCase::new(link_label.as_ref().into())) {
+                                    // found a matching definition!
+                                    let title = matching_def.title.as_ref().cloned().unwrap_or("".into());
+                                    let url = matching_def.dest.clone();
+                                    Some((link_type, url, title))
+                                } else if let Some(callback) = self.broken_link_callback {
+                                    // looked for matching definition, but didn't find it. try to fix
+                                    // link with callback, if it is defined
+                                    if let Some(val) = callback(link_label.as_ref(), link_label.as_ref()) {
```

Comment:
This would probably be clearer as `(name, title)` rather than `val`, so `name.into()` instead of `val.0.into()`.

Reply:
I don't think the strings need to separate from `&'a` when we annotate the return type with the same lifetime? E.g.,
```rust
broken_link_callback: Option<&'a Fn(&'a str) -> Option<(Cow<'a, str>, String)>>
```
Fully agreed on your second comment. Will change that.

Reply:
Thinking about it a bit more, I'm starting to question if `Cow`-like return types would even make sense here. The closure is passed a label, not an url, so it's not natural to expect it to be returned as either the url or title...

Reply:
Unifying lifetimes can cause borrow-checking problems for callers. This type of problem is subtle, so I'd only want to do it if we had evidence it was ok. Also what @marcusklaas said. I vote we keep it `String` and then possibly revisit this later.

Reply:
Ok! I'll fix up the other thing and merge it :-)